hls4ml/model/optimizer/passes/repack_stream.py

```diff
@@ -121,3 +122,14 @@ def transform(self, model, node):
         node.inputs[idx] = brdcst_out
 
         return True
+
+class RemoveFinalReshape(OptimizerPass):
+    ''' Remove reshape if final layer '''
+    def match(self, node):
+        # match if reshape is final node
+        return node.__class__.__name__ == 'Reshape' and not node.get_output_nodes()
+
+    def transform(self, model, node):
+        # remove, but don't rewire because it's the output layer
+        model.remove_node(node, rewire=False) 
```

Comment:
I don't think we should always do this, `io_stream` should still reshape the tensor if the model explicitly requested it.

Reply:
I think the idea is that when the `Reshape` is the output layer, it basically has no effect since the output is just a stream, which doesn't carry any inherent 'shape' information. But having the layer in seemed to cause the problem in #415

Reply:
See the test in #444 which passes for `io_stream` whether or not there is a `Reshape` (`repack_stream`) as the final operation (executed on master vs this PR, or only on this PR skipping the new optimizer)

Reply:
We can't say output is "just a stream", streams hold packed data (along the innermost dimension). If we always discard the last Reshape layer, we effectively lose the option of controlling the output of the IP.

Reply:
Okay I take the point about the packed data - so the `Reshape` is only a `nop` if the innermost dimension size doesn't change, right? But I do feel like controlling the model output dataflow with a `Reshape` layer is not a desirable feature

Reply:
It is more desirable that the tool does what it is expected as opposed to silently ignoring the user's wishes without even telling them "you don't want this, we know better than you".

Reply:
Sure, but is it really reasonable to expect that something like reshaping `(3, 3, 2) -> (18)` would change the dataflow? It's true for Vivado backend io_stream, but not io_parallel, and generally would depend on the backend implementation. This doesn't have to be silent either, we could add a warning in the printout, and optimizer passes can be skipped.

Reply:
Also 

> "you don't want this, we know better than you".

Isn't this an alternative definition of optimizer passes? 😇

Reply:
Modified as discussed offline to just remove the final reshape layer for `io_parallel` and warn for both `io_parallel` and `io_stream`